File: ml-benchmark/synduce/list/prodhom.f
Config NonLinear = true;

Inductive List = nil Unit | cons {Int, List};
Inductive CList = cnil Unit | single Int | concat {CList, CList};

max = \a: Int. \b: Int. if < a b then b else a;

spec = fix (
  \f: List -> Int. \xs: List.
  match xs with
    nil _ -> 1
  | cons {hd, tl} -> * hd (f tl)
  end
);

cat = fix (
  \f: List -> List -> List. \xs: List. \ys: List.
  match xs with
    nil _ -> ys
  | cons {a, b} -> cons {a, f b ys}
  end
);

repr = fix (
  \f: CList -> Compress List. \xs: CList.
  match xs with
    cnil _ -> nil unit
  | single a -> cons {a, nil unit}
  | concat {a, b} -> cat (f a) (f b)
  end
);

main = \x: CList. spec (repr x);

File: ml-benchmark/synduce/list/prodhom.ml
type list_ = Nil of unit | Cons of int * list_
type clist = CNil of unit | Single of int | Concat of clist * clist

let max a b = if a < b then b else a

let rec spec xs =
  match xs with
  | Nil _ -> 1
  | Cons (hd, tl) -> hd * spec tl

let rec cat xs ys =
  match xs with
  | Nil _ -> ys
  | Cons (a, b) -> Cons (a, cat b ys)

let rec repr xs =
  match xs with
  | CNil _ -> Nil ()
  | Single a -> Cons (a, Nil ())
  | Concat (a, b) -> cat (repr a) (repr b)

let program x = spec (repr x)